crates/p4-analyzer/src/native_fs.rs

```diff
@@ -0,0 +1,173 @@
+			async fn file_contents(file_uri: Url) -> Option<String> {
+				let path = file_uri.to_file_path().unwrap();
+
+				let data = tokio::fs::read(path).await;
+
+				match data {
+					Ok(data) => Some(String::from_utf8(data).unwrap()),
+					Err(_) => None,
+				}
+			}
+
+			Box::pin(file_contents(file_uri))
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::native_fs::NativeFs;
+	use analyzer_abstractions::lsp_types::Url;
+	use analyzer_host::lsp::RELATIVE_P4_SOURCEFILES_GLOBPATTERN;
+	use std::{fs, path::PathBuf};
+
+	#[tokio::test]
+	async fn test_enumerate_folder() {
+		// build test file_name location
+		let dir_name = "./test_directory0";
+		fs::create_dir(dir_name).expect("Failed to create directory");
+		fs::File::create("./test_directory0/file1.p4").unwrap();
+		fs::File::create("./test_directory0/file2.not_p4").unwrap();
+		fs::File::create("./test_directory0/file3.p4").unwrap();
+		fs::File::create("./test_directory0/file4.p4").unwrap();
+
+		// build NativeFs
+		let object = NativeFs::new();
+		let url = Url::from_directory_path(fs::canonicalize(&PathBuf::from(dir_name)).unwrap()).unwrap();
+
+		// tests
+		let result = object.enumerate_folder(url.clone(), RELATIVE_P4_SOURCEFILES_GLOBPATTERN.into()).await;
+
+		// Number of results
+		assert_eq!(result.len(), 3);
+		// URL content
+		assert_eq!(result[0].uri.scheme(), "file");
+		assert_eq!(result[0].uri.to_file_path().unwrap().file_name().unwrap(), "file1.p4");
+		assert_eq!(result[1].uri.scheme(), "file");
+		assert_eq!(result[1].uri.to_file_path().unwrap().file_name().unwrap(), "file3.p4");
+		assert_eq!(result[2].uri.scheme(), "file");
+		assert_eq!(result[2].uri.to_file_path().unwrap().file_name().unwrap(), "file4.p4");
+
+		// test recursive files & the OS file_name system has changes, so enumerate_folder() should give different result
+		fs::create_dir("./test_directory0/subdirectory0").expect("Failed to create directory");
+		fs::File::create("./test_directory0/subdirectory0/file5.p4").unwrap();
+		fs::File::create("./test_directory0/subdirectory0/file6.not_p4").unwrap();
+		fs::File::create("./test_directory0/subdirectory0/file7.p4").unwrap();
+		let result = object.enumerate_folder(url.clone(), RELATIVE_P4_SOURCEFILES_GLOBPATTERN.into()).await;
+
+		// Number of results
+		assert_eq!(result.len(), 5);
+		// URL content
+		assert_eq!(result[3].uri.scheme(), "file");
+		assert_eq!(result[3].uri.to_file_path().unwrap().file_name().unwrap(), "file5.p4");
+		assert!(result[3].uri.to_file_path().unwrap().parent().unwrap().to_str().unwrap().ends_with("subdirectory0"));
+		assert_eq!(result[4].uri.scheme(), "file");
+		assert_eq!(result[4].uri.to_file_path().unwrap().file_name().unwrap(), "file7.p4");
+		assert!(result[4].uri.to_file_path().unwrap().parent().unwrap().to_str().unwrap().ends_with("subdirectory0"));
+
+		// clean up
+		fs::remove_dir_all(dir_name).expect("Failed to delete directory");
```

Comment:
If an assertion fails (or anything else causing a panic) in the line above, then this cleanup code won't be executed. This will leave the `'test_directory0'` folder hanging around and subsequent runs of the test fail because the folder already exists.